### PR TITLE
Deprecated Agent State to Agent Phase

### PR DESCRIPTION
**Pull request by Future-Outlier**
## Tracking issue
Fixes https://github.com/flyteorg/flyte/issues/3936

## Why are the changes needed?
We want to deprecate `State` in webapi and use `core.TaskExecution.Phase` instead to make users less confused.

## What changes were proposed in this pull request?
- add `core.TaskExecution.Phase` phase in `agent.proto`
- add `DEPRECATED` annotation in `State` in `agent.proto`
- add switch case statements to route the phase in `agent/plugin.go`
- add tests for above

Note:
The reason why we need to:
1. check the phase
2. check the state

is because the default value of phase is `UNDEFINED`, and the default value of state is `RETRYABLE FAILURE`.
We can check the state when the phase is `UNDEFINED`.

## How was this patch tested?
- single binary and flytekit [here](https://github.com/flyteorg/flytekit/pull/2123)
- unit test
### Setup process
0. start agent server by flytekit
1. start flyte components with `--dev`
2. compile and start the flyte cluster
3. run a remote task with sensor agent

```python
pyflyte serve agent
```
```bash
flytectl demo start --dev
```
```bash
cd flyte
make compile;
flyte start --config ../flyte-single-binary-local-dev.yaml
```
```bash
pyflyte run --remote sensor_example.py wf
```

```yaml
tasks:
  task-plugins:
    enabled-plugins:
      - agent-service

plugins:
  agent-service:
    supportedTaskTypes:
      - sensor
    defaultAgent:
      endpoint: "dns:///localhost:8000"
      insecure: true
      timeouts:
        GetTask: 100s
      defaultTimeout: 100s
```
```python
from flytekit.sensor.file_sensor import FileSensor
from flytekit import ImageSpec, task, workflow
sensor = FileSensor(name="test_sensor")

@task()
def t1():
    return

@workflow()
def wf():
    sensor(path="/tmp") >> t1()

if __name__ == "__main__":
    wf()
```

### Screenshots
image

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
https://github.com/flyteorg/flytekit/pull/2123

### CODE REVIEW

**Comment by Future-Outlier** [comment]
image

File: flyteidl/protos/flyteidl/admin/agent.proto [hunk 1]
```diff
@@ -78,7 +78,7 @@ message GetTaskResponse {
 }
 
 message Resource {
-  // The state of the execution is used to control its visibility in the UI/CLI.
+  // DEPRECATED. The state of the execution is used to control its visibility in the UI/CLI.
```
Comment (honnix) [inline flyteidl/protos/flyteidl/admin/agent.proto:81]: There is a `deprecated` keyword to use: https://protobuf.dev/programming-guides/proto3/#options

```
State state = 1 [deprecated = true];
```

File: flyteplugins/go/tasks/plugins/webapi/agent/plugin.go [hunk 2]
```diff
@@ -190,6 +193,31 @@ func (p Plugin) Status(ctx context.Context, taskCtx webapi.StatusContext) (phase
...
+			return core.PhaseInfoUndefined, err
+		}
+		return core.PhaseInfoSuccess(taskInfo), nil
+	case flyteIdl.TaskExecution_ABORTED:
+		return core.PhaseInfoFailure(pluginErrors.TaskFailedWithError, "failed to run the job with aborted phase", taskInfo), nil
+	case flyteIdl.TaskExecution_FAILED:
+		return core.PhaseInfoFailure(pluginErrors.TaskFailedWithError, "failed to run the job", taskInfo), nil
+	}
+
+	logger.Infof(ctx, "Agent returned an undefined phase [%v]. Checking state.", resource.Phase)
```
Comment (honnix) [inline flyteplugins/go/tasks/plugins/webapi/agent/plugin.go:219]: I'm not sure I like this type of fallback logic. Suppose we add a new phase later, the previous switch would not match that and then we fallback to state, which is obviously not what we want.

I prefer explicitly checking whether `resource.Phase` is `UNDEFINED` and then use state.

Comment (Future-Outlier) [reply flyteplugins/go/tasks/plugins/webapi/agent/plugin.go:219]: This is a great idea! Thank you so much!

File: flyteidl/protos/flyteidl/admin/agent.proto (hunk 1 above)
Comment (Future-Outlier) [reply flyteidl/protos/flyteidl/admin/agent.proto:81]: Thank you so much! I learned a lot

Comment (Future-Outlier) [reply flyteidl/protos/flyteidl/admin/agent.proto:81]: Fixed! Thank you again!

File: flyteplugins/go/tasks/plugins/webapi/agent/plugin.go (hunk 2 above)
Comment (Future-Outlier) [reply flyteplugins/go/tasks/plugins/webapi/agent/plugin.go]: Fixed! Thank you again!

**Review by pingsutw — APPROVED** [review]

**Pull request merged**